### PR TITLE
Smart Health Cards Revocations

### PR DESCRIPTION
**Pull request by JosephFreshworks**
SMART Health Card has implemented a new feature called a [revocation ID] (a unique identifier for each PVC). The unique identifier will allow provinces and territories to de-authorize (or revoke) select PVCs (e.g., for fraudulent reasons). The rid values may also be timeboxed. Similar to the approved issuers, each province/territory will have a list of unique identifiers (new endpoint) that will read as ‘Does not meet requirements'.

https://freshworks.atlassian.net/browse/YKVERIFIER-116

### CODE REVIEW

File: shcDecoder/src/main/java/ca/bc/gov/shcdecoder/revocations/Utils.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package ca.bc.gov.shcdecoder.revocations
+
+import ca.bc.gov.shcdecoder.cache.impl.CacheManagerImpl
+
+private const val REVOCATION_JSON_PATH = "/.well-known/crl/"
+
+fun getRevocationsUrl(iss: String, kid: String): String {
```
Comment (pinakinkansara) [inline shcDecoder/src/main/java/ca/bc/gov/shcdecoder/revocations/Utils.kt:7]: we can have a private method within the Revocation ManagerImpl class.
As this is not reusable method for any other area of our code

File: shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/FileManagerImpl.kt [hunk 2]
```diff
@@ -6,16 +6,20 @@ import ca.bc.gov.shcdecoder.cache.FileManager
...
 import com.google.gson.Gson
 import java.io.File
 import java.io.FileOutputStream
 import java.net.URL
+import java.util.Date
 
 class FileManagerImpl(
-    context: Context
+    context: Context,
+    private val gson: Gson
```
Comment (pinakinkansara) [inline shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/FileManagerImpl.kt:22]: Please revert this variable declaration to method scope.
with refactored code, I don't see its usage beyond one function.

File: shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/FileManagerImpl.kt [hunk 3]
```diff
@@ -59,29 +63,51 @@ class FileManagerImpl(
...
+        val validationRulesResponse = getDataFromFile(url, ValidationRuleResponse::class.java)
+        return validationRulesResponse?.ruleSet.orEmpty()
     }
 
-    private fun getJsonStringFromFile(url: String): String {
-        val fileName = getFileNameFromUrl(url)
-        val file = File(downloadDir, fileName)
-        val bufferedReader = file.bufferedReader()
-        return bufferedReader.use { it.readText() }
+    override suspend fun getRevocations(url: String): List<Pair<String, Date?>> {
```
Comment (pinakinkansara) [inline shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/FileManagerImpl.kt:80]: getRecovations & getRevocationsCtr can be clubbed in to one method with a custom model class that hold both this return type in single data class.

File: shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/CacheManagerImpl.kt [hunk 4]
```diff
@@ -17,27 +19,50 @@ internal class CacheManagerImpl(
...
-                val issuers = fileManager.getIssuers(shcConfig.issuerEndPoint)
-                issuers.forEach { issuer ->
-                    val keyUrl = if (issuer.iss.endsWith(SUFFIX_JWKS_JSON)) {
-                        issuer.iss
-                    } else {
-                        "${issuer.iss}$SUFFIX_JWKS_JSON"
-                    }
-                    fileManager.downloadFile(keyUrl)
-                }
+            fileManager.downloadFile(shcConfig.rulesEndPoint)
```
Comment (pinakinkansara) [inline shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/CacheManagerImpl.kt:26]: Please wrap this inside try-catch block as we don't want to block download & cache for other URL if one of them fails.

**Review by pinakinkansara — CHANGES_REQUESTED** [review]
I have added code review comments.
Set up a meeting if you need any clarification on that.

File: shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/CacheManagerImpl.kt (hunk 4 above)
Comment (JosephFreshworks) [reply shcDecoder/src/main/java/ca/bc/gov/shcdecoder/cache/impl/CacheManagerImpl.kt:26]: the downloadFile method code is wrapped in a try catch block, in theory it should not be possible for this to block others. What do you think?

**Review by pinakinkansara — APPROVED** [review]

**Pull request merged**